File: 백준/Silver/4889. 안정적인 문자열/안정적인 문자열.java
import java.util.*;
import java.lang.*;
import java.io.*;

class Main {
    static Stack<Character> st = new Stack<>();
    static int cnt = 1;
	public static void main (String[] args) throws java.lang.Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        
        while(true){
            String str = br.readLine();
            
            if(str.charAt(0) == '-')
                break;
            
            for(int i = 0; i < str.length(); i++){
                if(st.isEmpty()){
                    st.add(str.charAt(i));
                }
                else{
                    if(str.charAt(i) == '}'){
                        if(st.peek() == '{'){
                            st.pop();
                        }
                        else{
                            st.add(str.charAt(i));
                        }
                    }
                    else
                        st.add('{');
                }
            }
            
            int res = 0;
            while(!st.isEmpty()){
                char b = st.pop();
                char a = st.pop();
                
                if(a == b){
                    res += 1;
                }
                else
                    res += 2;
            }
            
            System.out.println(cnt + ". " + res);
            cnt++;
            st.clear();
        }
        
	}
}

File: 백준/Silver/4889. 안정적인 문자열/안정적인 문자열.cc
#include <iostream>
#include <stack>
#include <string>

using namespace std;

int main() {
    cin.tie(NULL);
    ios::sync_with_stdio(false);
    string str;
    int cnt = 1;
    while(true){
        cin >> str;
        
        if(str[0] == '-'){
            break;
        }
        
        stack<char> st;
        
        for(int i = 0; i < str.length(); i++){
            if(st.empty()){
                st.push(str[i]);
            }        
            else{
                if(str[i] == '}'){
                    if(st.top() == '{')
                        st.pop();
                    else
                        st.push(str[i]);
                }
                else{
                    st.push({str[i]});
                }
            }
        }
        int res = 0;
        
        while(!st.empty()){
            char b = st.top();
            st.pop();
            char a = st.top();
            st.pop();
            
            if(a == b){
                res += 1;
            }
            else{
                res += 2;
            }
        }
    
        cout << cnt << ". " << res << '\n';
        cnt++;
    }
    return 0;
}